utils/parametric-components.stanza

```diff
@@ -0,0 +1,537 @@
+            fatal("Instance %_ does not have expected property (user-params)" % [ref(my-inst)])
+          val user-params = replace-param(property(my-inst.user-params))
+          ; println("Searching with user-params => %_" % [user-params])
+          val new-comp =
+            if has-property?(my-inst.resistor) :
+              try :
+                chip-resistor(user-params)
+              catch (e):
+                println("Warning: generic resistor substitution for %_" % [ref(my-inst)])
+                match(property(my-inst.resistance)) :
+                  (r:Double) : dummy-resistor(r)
+                  (t:Toleranced) : dummy-resistor(find-std-val(t), t)
+            else if has-property?(my-inst.capacitor) :
+              try :
+                ceramic-cap(user-params)
+              catch (e):
+                println("Warning: generic capacitor substitution for %_" % [ref(my-inst)])
+                match(property(my-inst.capacitance)) :
+                  (c:Double) : dummy-capacitor(c)
+                  (t:Toleranced) : dummy-capacitor(find-std-val(t), t)
+            else if has-property?(my-inst.inductor) :
+              try :
+                smd-inductor(user-params)
+              catch (e):
+                println("Warning: generic inductor substitution for %_" % [ref(my-inst)])
+                match(property(my-inst.inductance)) :
+                  (l:Double) : dummy-inductor(l)
+                  (t:Toleranced) : dummy-inductor(find-std-val(t), t)
+            else :
+              fatal("unknown component found")
+          inst my-new-comp : new-comp
+          property(my-new-comp.name) = to-symbol(ref(my-inst))
+          ; hook in new component to old parametric instance pins (in parallel)
+          net (my-new-comp.p[1] my-inst.p[1])
+          net (my-new-comp.p[2] my-inst.p[2])
+          add(new-comps, new-comp)
+    new-comps
+
+  within transform-module(m) :
+    inside pcb-module :
+      val knownR = HashTable<Double|Toleranced, Vector<JITXObject>>()
+      val knownC = HashTable<Double|Toleranced, Vector<JITXObject>>()
+      val knownL = HashTable<Double|Toleranced, Vector<JITXObject>>()
+      val paraR = HashTable<Double|Toleranced, Vector<JITXObject>>()
+      val paraC = HashTable<Double|Toleranced, Vector<JITXObject>>()
+      val paraL = HashTable<Double|Toleranced, Vector<JITXObject>>()
+
+      ; Need to recursively retrieve all instances (including instances of modules.)
+      ; Loop runs twice: 1 find, 1 add part.
+      for i in list-all-instances() do :
+        if has-property?(i.parametric) :
+          ; println("Sorting part %_ with desired value %_" % [ref(i) property(i.resistance)])
+          if has-property?(i.resistor) :
+            match(get?(paraR, property(i.resistance))) : 
+              (v:Vector<JITXObject>) : add(v,i)
+              (_) : 
+                val t = Vector<JITXObject>()
+                add(t,i)
+                set(paraR, property(i.resistance), t)
```

Comment:
```
if has-property?(i.resistor) :
  add(paraR[property(i.resistance)], t)
```
after using `HashTable-init`

Reply:
Very cool.

Reply:
So I don't quite understand how to integrate your code.
```
      for i in list-all-instances() do :
        if has-property?(i.parametric) :
          ; println("Sorting part %_ with desired value %_" % [ref(i) property(i.resistance)])
          if has-property?(i.resistor) :
            add(paraR[property(i.resistance)], t)
```
where does the `t` come from? If the key (i.resistance) doesn't exist the old code created `t` as a vector which became the value for the hashtable.

Reply:
Sorry, `i` not `t`.
```
if has-property?(i.resistor) :
  add(paraR[property(i.resistance)], i)
```

Reply:
Just assume there is already a vector in `paraR[key]`, for any `res`. So you can safely append to it.

Reply:
Cool.